### PR TITLE
[backport] SI-7470 implements fundep materialization

### PR DESCRIPTION
**Pull request by xeno-by**
Backports 21a8c6c from the 2.11.x branch under -Xfundep-materialization
as per Miles Sabin's request. Thanks Miles!

### CODE REVIEW

**Comment by xeno-by** [comment]
review @adriaanm @retronym @milessabin

**Comment by xeno-by** [comment]
The original pull request for 2.11.x: https://github.com/scala/scala/pull/2821

**Comment by adriaanm** [comment]
Thanks, thinking out loud: would it make sense to use -Xsource:2.11 for this instead. It doesn't really matter for 2.10, but in 2.11 we could centralize these things this way.

**Comment by milessabin** [comment]
@xeno-by awesome ... thanks!

@adriaanm -Xsource:2.11 would work just as well for me, but what about -Xfuture?

**Comment by adriaanm** [comment]
I'd like to subsume -Xfuture by -Xsource:$currentVersion.next (because we already do the latter for $currentVersion.prev)

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 1]
```diff
@@ -886,10 +908,12 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
    *    2) undetparams (sym.isTypeParameter && !sym.isSkolem)
    */
   var hasPendingMacroExpansions = false
+  private val forced = perRunCaches.newWeakSet[Tree]
   private val delayed = perRunCaches.newWeakMap[Tree, scala.collection.mutable.Set[Int]]
   private def isDelayed(expandee: Tree) = delayed contains expandee
   private def calculateUndetparams(expandee: Tree): scala.collection.mutable.Set[Int] =
-    delayed.get(expandee).getOrElse {
+    if (forced(expandee)) scala.collection.mutable.Set[Int]()
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:916]: please add a comment: `!settings.XfundepMaterialization.value implies forced.isEmpty`

**Comment by adriaanm** [comment]
Whatever we decided, I'd rather not introduce another -X setting. I'd prefer -Xsource, but -Xfuture is fine too. (I forget whether 2.10 already has -Xsource.)

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:916]: done

**Comment by xeno-by** [comment]
I don't mind using any of the flags. Speaking of -Xsource:2.11, is there any precedent of it being used for something else?

**Comment by adriaanm** [comment]
It's used in the 2.11 codebase to be backwards compatible with 2.10.

**Comment by xeno-by** [comment]
Yeah, I saw that -Xsource:2.10 is used, but what about -Xsource:2.11? Is the particular 2.11 value used anywhere?

**Comment by adriaanm** [comment]
I don't think so. We should systematically use it in 2.12 for any source
backwards incompatible changes -- conversely for 2.10

**Comment by xeno-by** [comment]
Allright, then we set the precedent and go for -Xsource:2.11 now?

**Comment by adriaanm** [comment]
I think that would be more consistent. I won't insist too much, since 2.10.5 will be the last 2.10 release anyway.

**Comment by xeno-by** [comment]
Funnily enough, we don't even have -Xsource in 2.10.x. How about we just use -Xfundep-materialization then? :)

**Comment by adriaanm** [comment]
Sure. I suspected as much, but didn't have time to check.

**Comment by gkossakowski** [comment]
@xeno-by, pinging on addressing @adriaanm's feedback. It seems like `-Xfundep-materialization` is the way to go, right?

**Comment by xeno-by** [comment]
Looks like it, yes. I think I've addressed everything btw.

**Comment by gkossakowski** [comment]
True. Thinking about it more, could we hide this behind 'Y' flag instead? We have lesser obligation to keep around `Y` flags and this is something we should remove from 2.11/2.12.

**Comment by xeno-by** [comment]
We could do that, yes. Shall I rename the flag then? Any other comments that I could fix while I'm at it?

**Comment by gkossakowski** [comment]
No, I think that's all I have. I don't understand the feature itself but I trust that you and @milessabin has spent enough of time on it :-)

Once you change the flag, ping me and I'll merge it.

**Comment by xeno-by** [comment]
Ping

**Comment by xeno-by** [comment]
Ping @gkossakowski. Shall we merge?

**Comment by gkossakowski** [comment]
Yep.

**Comment by milessabin** [comment]
Thanks.

**Pull request merged**